test/zero-downtime/testsuite_zerodowntime_test.go

```diff
@@ -0,0 +1,176 @@
+package zero_downtime
+
+import (
+	"fmt"
+	testutils "github.com/keptn/keptn/test/go-tests"
+	"github.com/stretchr/testify/suite"
+	"os"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+const EnvInstallVersion = "INSTALL_HELM_CHART"
+const EnvUpgradeVersion = "UPGRADE_HELM_CHART"
+const valuesFile = "test-values.yaml"
+
+const apiProbeInterval = 5 * time.Second
+const sequencesInterval = 15 * time.Second
+
+type ZeroDowntimeEnv struct {
+	quit         chan struct{}
+	NrOfUpgrades int
+	Wg           *sync.WaitGroup
+
+	//api test fields
+	TotalAPICalls  uint64
+	FailedAPICalls uint64
+	PassedAPICalls uint64
+
+	//sequence related test fields
+	ShipyardFile    string
+	ExistingProject string
+	FiredSequences  uint64
+	FailedSequences uint64
+	PassedSequences uint64
+	Id              uint64
+}
+
+func SetupZD() *ZeroDowntimeEnv {
+
+	zd := ZeroDowntimeEnv{}
+	zd.quit = make(chan struct{})
+	zd.NrOfUpgrades = 2
+	zd.Wg = &sync.WaitGroup{}
+	zd.ShipyardFile, _ = GetShipyard()
+	zd.TotalAPICalls = 0
+	zd.FailedAPICalls = 0
+	zd.PassedAPICalls = 0
+	zd.FiredSequences = 0
+	zd.FailedSequences = 0
+	zd.PassedSequences = 0
+	zd.Id = 0
+	return &zd
+}
+
+func (env *ZeroDowntimeEnv) gedId() string {
+	atomic.AddUint64(&env.Id, 1)
+	return fmt.Sprintf("%d", env.Id)
+}
+
+func (env *ZeroDowntimeEnv) failSequence() {
+	atomic.AddUint64(&env.FailedSequences, 1)
+}
+
+func (env *ZeroDowntimeEnv) passSequence() {
+	atomic.AddUint64(&env.PassedSequences, 1)
+}
+
+func (env *ZeroDowntimeEnv) passFailedSequence() {
+	atomic.AddUint64(&env.FailedSequences, ^uint64(1-1))
+	env.passSequence()
+}
+
+type TestSuiteDowntime struct {
+	suite.Suite
+}
+
+func (suite *TestSuiteDowntime) SetupSuite() {
+
+}
+
+//Test_ZeroDowntime runs all test suites
+func Test_ZeroDowntime(t *testing.T) {
+	suite.Run(t, new(TestSuiteDowntime))
+}
+
+func (suite *TestSuiteDowntime) TestSequences() {
+	ZDTestTemplate(suite.T(), Sequences, "Sequences")
+}
+
+func (suite *TestSuiteDowntime) TestWebhook() {
+	ZDTestTemplate(suite.T(), Webhook, "Webhook")
+}
+
+func (suite *TestSuiteDowntime) TearDownSuite() {
+}
+
+func ZDTestTemplate(t *testing.T, F func(t1 *testing.T, e *ZeroDowntimeEnv), name string) {
+
+	env := SetupZD()
+	t.Run("Rolling Upgrade", func(t2 *testing.T) {
+		t2.Parallel()
+		RollingUpgrade(t2, env)
+	})
+
+	t.Run("API", func(t2 *testing.T) {
+		t2.Parallel()
+		APIs(t2, env)
+	})
+
+	t.Run(name, func(t2 *testing.T) {
+		t2.Parallel()
+		F(t2, env)
+
+	})
+
+}
+
+func RollingUpgrade(t *testing.T, env *ZeroDowntimeEnv) {
+	defer func() {
+		close(env.quit)
+		t.Log("Rolling upgrade terminated")
+		env.Wg.Wait()
+
+		PrintSequencesResults(t, env)
```

Comment:
would it make sense to call the `PrintResults()` methods of the other tests in `ZDTestTemplate()`? This might be a bit cleaner, as the `RollingUpgrade()` would then be independent from the tests that are executed during the upgrades

Reply:
the problem is that since the template is parallel the code would be run immediately and sometimes there is a race condition for which the sequence test has not yet added to the wait group, so we would print before the end of the test... I could try to add a sleep there maybe ...

ex. run 
![image](https://user-images.githubusercontent.com/89971034/166411941-9a8e0c96-e2e4-4478-802f-474439a48bab.png)

Reply:
adding  wait in the templates is blocking 😢